src/modules/csm/state.py

```diff
@@ -89,22 +95,58 @@ def file(cls) -> Path:
     def buffer(self) -> Path:
         return self.file().with_suffix(".buf")
 
+    @property
+    def is_empty(self) -> bool:
+        return not self.data and not self._epochs_to_process and not self._processed_epochs
+
+    @property
+    def unprocessed_epochs(self) -> set[EpochNumber]:
+        if not self._epochs_to_process:
+            raise ValueError("Epochs to process are not set")
+        diff = set(self._epochs_to_process) - self._processed_epochs
+        return diff
+
+    @property
+    def is_fulfilled(self) -> bool:
+        return not self.unprocessed_epochs
+
+    @property
+    def frames(self):
+        return self.calculate_frames(self._epochs_to_process, self._epochs_per_frame)
+
+    @staticmethod
+    def calculate_frames(epochs_to_process: tuple[EpochNumber, ...], epochs_per_frame: int) -> list[Frame]:
+        """Split epochs to process into frames of `epochs_per_frame` length"""
+        if len(epochs_to_process) % epochs_per_frame != 0:
+            raise ValueError("Insufficient epochs to form a frame")
+        return [(frame[0], frame[-1]) for frame in batched(epochs_to_process, epochs_per_frame)]
+
     def clear(self) -> None:
-        self.data = defaultdict(AttestationsAccumulator)
+        self.data = {}
         self._epochs_to_process = tuple()
         self._processed_epochs.clear()
         assert self.is_empty
 
-    def inc(self, key: ValidatorIndex, included: bool) -> None:
-        self.data[key].add_duty(included)
+    def find_frame(self, epoch: EpochNumber) -> Frame:
+        for epoch_range in self.frames:
+            if epoch_range[0] <= epoch <= epoch_range[1]:
+                return epoch_range
+        raise ValueError(f"Epoch {epoch} is out of frames range: {self.frames}")
+
+    def increment_duty(self, frame: Frame, val_index: ValidatorIndex, included: bool) -> None:
+        if frame not in self.data:
+            raise ValueError(f"Frame {frame} is not found in the state")
+        self.data[frame][val_index].add_duty(included)
 
     def add_processed_epoch(self, epoch: EpochNumber) -> None:
         self._processed_epochs.add(epoch)
 
     def log_progress(self) -> None:
         logger.info({"msg": f"Processed {len(self._processed_epochs)} of {len(self._epochs_to_process)} epochs"})
 
-    def migrate(self, l_epoch: EpochNumber, r_epoch: EpochNumber, consensus_version: int):
+    def init_or_migrate(
```

Comment:
Still think it's just `migrate`)

Reply:
arguable, because we do nothing meaningful in `State.__init__`. from my perspective we do init when fill `epochs_to_process` with value and other things as well

Reply:
I think you speak of different things. `__init__` is for an instantiation, i.e. creating an instance. It can accept whatever you need to construct the object, let's say maximum expected size of some array, and do not accept any argument to fill in data in the object, e.g., you are creating a connection to a database.

That's why I think `migrate` is suitable here. State object is "initialized" upon instantiation and its API says, use `migrate` whenever you want to work with some epochs range, and the object will do whatever it needs, including filling in missing variables, preserving the accumulated data, and so on.